src/credentials/CHIPCert.h

```diff
@@ -0,0 +1,460 @@
+    void Release(void);
+    void Clear(void);
+
+    CHIP_ERROR LoadCert(const uint8_t * chipCert, uint32_t chipCertLen, uint16_t decodeFlags, ChipCertificateData *& cert);
+    CHIP_ERROR LoadCert(TLVReader & reader, uint16_t decodeFlags, ChipCertificateData *& cert);
+    CHIP_ERROR LoadCerts(const uint8_t * chipCerts, uint32_t chipCertsLen, uint16_t decodeFlags);
+    CHIP_ERROR LoadCerts(TLVReader & reader, uint16_t decodeFlags);
+    CHIP_ERROR AddTrustedKey(uint64_t caId, OID curveOID, const uint8_t * pubKey, uint16_t pubKeyLen, const uint8_t * pubKeyId,
+                             uint16_t pubKeyIdLen);
+    ChipCertificateData * FindCert(const CertificateKeyId & subjectKeyId) const;
+    ChipCertificateData * LastCert(void) const { return (CertCount > 0) ? &Certs[CertCount - 1] : nullptr; }
+
+    CHIP_ERROR ValidateCert(ChipCertificateData & cert, ValidationContext & context);
+    CHIP_ERROR FindValidCert(const ChipDN & subjectDN, const CertificateKeyId & subjectKeyId, ValidationContext & context,
+                             ChipCertificateData *& cert);
+
+    CHIP_ERROR VerifySignature(ChipCertificateData & cert, ChipCertificateData & caCert);
+
+protected:
+    AllocFunct mAllocFunct;
+    FreeFunct mFreeFunct;
+    uint8_t * mDecodeBuf;
+    uint16_t mDecodeBufSize;
+
+    CHIP_ERROR FindValidCert(const ChipDN & subjectDN, const CertificateKeyId & subjectKeyId, ValidationContext & context,
+                             uint16_t validateFlags, uint8_t depth, ChipCertificateData *& cert);
+    CHIP_ERROR ValidateCert(ChipCertificateData & cert, ValidationContext & context, uint16_t validateFlags, uint8_t depth);
+};
+
+extern CHIP_ERROR DecodeChipCert(const uint8_t * chipCert, uint32_t chipCertLen, ChipCertificateData & certData);
+extern CHIP_ERROR DecodeChipCert(TLVReader & reader, ChipCertificateData & certData);
+
+extern CHIP_ERROR DecodeChipDN(TLVReader & reader, ChipDN & dn);
+
+extern CHIP_ERROR ConvertX509CertToChipCert(const uint8_t * x509Cert, uint32_t x509CertLen, uint8_t * chipCertBuf,
+                                            uint32_t chipCertBufSize, uint32_t & chipCertLen);
+extern CHIP_ERROR ConvertChipCertToX509Cert(const uint8_t * chipCert, uint32_t chipCertLen, uint8_t * x509CertBuf,
+                                            uint32_t x509CertBufSize, uint32_t & x509CertLen);
+
+extern CHIP_ERROR DetermineCertType(ChipCertificateData & cert);
+
+extern CHIP_ERROR PackCertTime(const chip::ASN1::ASN1UniversalTime & time, uint32_t & packedTime);
+extern CHIP_ERROR UnpackCertTime(uint32_t packedTime, chip::ASN1::ASN1UniversalTime & time);
+extern uint16_t PackedCertTimeToDate(uint32_t packedTime);
+extern uint32_t PackedCertDateToTime(uint16_t packedDate);
+extern uint32_t SecondsSinceEpochToPackedCertTime(uint32_t secondsSinceEpoch);
+
+inline void ValidationContext::Reset()
+{
+    memset((void *) this, 0, sizeof(*this));
```

Comment:
replace c-style cast with C++ cast. I am not a fan of memset(this) in C++... could we do data member reset to make it clear what gets set?

Without it, enums just get 0 without actually saying what those values mean.